### PR TITLE
Improve plugin resolution (fixes pnpm compatibility)

### PR DESCRIPTION
**Pull request by adamhl8**
## Description

Fixes https://github.com/prettier/prettier/issues/8056

**== See this reply for an explanation of the updated fix: https://github.com/prettier/prettier/pull/11248#issuecomment-886130669 ==**

~~Previously, plugins were loaded from the `node_modules` directory that was parent to the current running instance of Prettier. Loading relative to the cwd ensures that the right `node_modules` is used (for example, when using Prettier with pnpm).~~

~~I added [pkg-dir](https://github.com/sindresorhus/pkg-dir) to handle finding the project directory. I initially wrote my own way to find it, but I figured using this package is probably better and safer than anything I could come up with. This package simply starts from the cwd and traverses up until it finds a path where `package.json` exists. Hopefully this wouldn't break plugin loading for anyone. I can't think of any project setups where this would.~~

~~I've tested this with my own projects that use pnpm and Prettier loads all my plugins without issue.~~

## Checklist

- [ ] I’ve added tests to confirm my change works.

I could be completely wrong, but I don't think additional tests need to be written for this change. All current tests pass. `cwd` will always be a path in the actual project that is being formatted, so the right `node_modules` will be found.

- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).

I can update the docs if needed, but it seems like the way `plugins.md` is currently worded still makes sense.

```
Providing at least one path to `--plugin-search-dir`/`pluginSearchDirs`
turns off plugin autoloading in the default directory (i.e. `node_modules` above `prettier` binary).
```

The default directory is still the `node_modules` above the `prettier` binary, we're just getting there a different way.

- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
Thanks for working on this.

1. We search project root differently in other place, see [this](https://github.com/prettier/prettier/blob/main/src/config/find-project-root.js)
2. If we decide to change search logic, what do you think just search from `process.cwd()`?
3. Big problem, we still support global installation, this solution will break that, I'm not sure what's right to do.

**Comment by fisker** [comment]
> This package simply starts from the cwd and traverses up until it finds a path where package.json exists. Hopefully this wouldn't break plugin loading for anyone. I can't think of any project setups where this would.

This will break monorepo.

**Comment by adamhl8** [comment]
I've come up with a solution that fixes compatibility with pnpm and shouldn't (hopefully) break any project setups (monorepo, global install, etc). It also doesn't care about what the virtual store is called or where it is.

I added tests and updated the plugins documentation page as well. I'm not very familiar with jest so I'd appreciate if someone could check those over.

There are comments in the code that hopefully makes it clear, but I'll somewhat briefly explain here:

Essentially, instead of trying to find pnpm's `node_modules` directory (which contains the symlinks), we just use the virtual store directly since this is where Prettier is being run from. (The virtual store contains hardlinks to the *actual* packages but that's irrelevant in this case.)

All packages in the virtual store have a directory structure that looks like this:

```
.pnpm/ (virtual store)
├── <package>@<version>/ (e.g. prettier@2.3.2)
│   └── node_modules/ (contains package + its dependencies)
│       ├── prettier/ (dir where prettier is run from)
│       │   ├── index.js
│       │   └── ...
│       ├── dependency1/
│       └── ...
├── prettier-plugin-foo@1.2.3
│   └── node_modules
│       └── prettier-plugin-foo
└── ...
```

All packages in the virtual store are appended with `@<version>`. We can check for that to determine if we're in the virtual store, move up into the parent directory (virtual store), and get the path for each plugin directory. Because each plugin directory will have its own `node_modules`, we can simply pass each path to the existing plugin resolution function.

Additionally, you can now pass `--plugin-search-dir` any arbitrary directory that contains Prettier plugins. This shouldn't affect anyone already loading plugins this way. i.e. you can pass `node_modules` directly or a directory that is the direct parent to plugins:

```
somePluginsDir/ (or node_modules)
├── prettier-plugin-foo/
└── prettier-plugin-bar/
```

**Comment by adamhl8** [comment]
@fisker Any thoughts on this?

**Comment by malekjaroslav** [comment]
@fisker @adamhl8 Any updates on this, it's a very annoying bug not being able to use Prettier plugins with pnpm

**Comment by adamhl8** [comment]
> It seems there are lots of unrelated changes.

What do you mean?

**Comment by adamhl8** [comment]
> It seems there are lots of unrelated changes.

Sorry for the slow response.
Unrelated to what? I feel like all the changes are within the "scope" of this PR. I realize that "Improve plugin resolution" is a bit broad, but that's what this PR aims to do. And a side effect of that being it fixes plugin compatibility with pnpm.

**Comment by rburgst** [comment]
any updates on this? I would love to be able to use pnpm

**Comment by Shinigami92** [comment]
> any updates on this? I would love to be able to use pnpm

What's stopping you?
Just configure it like in this project: https://github.com/prettier/plugin-pug/blob/3c6dd7222b65fea5c2ba69ddbb00dbed56d70b86/.prettierrc.cjs#L8

**Comment by ZeroAurora** [comment]
> > any updates on this? I would love to be able to use pnpm
> 
> What's stopping you? Just configure it like in this project: https://github.com/prettier/plugin-pug/blob/3c6dd7222b65fea5c2ba69ddbb00dbed56d70b86/.prettierrc.cjs#L8

But it's something that should work out of box instead of manual configuring. Not to mention those who prefer to use JSON as the configuration format (like me).

**Comment by TicTak21** [comment]
Any updates?

**Comment by adamhl8** [comment]
I'd appreciate if I could get some more detailed feedback on what is preventing this PR from moving forward. @fisker

**Comment by fisker** [comment]
Can you explain to me how this change https://github.com/prettier/prettier/pull/11248/files#r724610453 is related?

**Comment by adamhl8** [comment]
I updated/cleaned up the plugins documentation to reflect the [changes](https://github.com/prettier/prettier/pull/11248#issuecomment-886130669) I made. I'm honestly not trying to be difficult; I don't understand how any of these changes are unrelated. I can remove the changes I made to the docs, but then the docs wouldn't accurately reflect what you can do with the plugin CLI flags or the config file.

File: docs/plugins.md [hunk 1]
```diff
@@ -3,35 +3,44 @@ id: plugins
 title: Plugins
 ---
 
-Plugins are ways of adding new languages to Prettier. Prettier’s own implementations of all languages are expressed using the plugin API. The core `prettier` package contains JavaScript and other web-focused languages built in. For additional languages you’ll need to install a plugin.
+Plugins are ways of adding new languages or formatting rules to Prettier. Prettier’s own implementations of all languages are expressed using the plugin API. The core `prettier` package contains JavaScript and other web-focused languages built in. For additional languages you’ll need to install a plugin.
 
 ## Using Plugins
 
-Plugins are automatically loaded if you have them installed in the same `node_modules` directory where `prettier` is located. Plugin package names must start with `@prettier/plugin-` or `prettier-plugin-` or `@<scope>/prettier-plugin-` to be registered.
+Plugins are automatically loaded if you have them installed in the same `node_modules` directory where `prettier` is located. Plugin package names must start with `prettier-plugin-` or `@prettier/plugin-` or `@<scope>/prettier-plugin-` to be registered.
```
[the remark this answers is not in the document]
Comment (adamhl8) [reply docs/plugins.md:10]: This is just a change of the ordering of the text to make things a bit more consistent throughout the docs.

File: docs/plugins.md [hunk 2]
```diff
@@ -3,35 +3,44 @@ id: plugins
...
 
-- Or the [API](api.md), via the `plugins` and `pluginSearchDirs` options:
+- The [API](api.md), via the `plugins` and `pluginSearchDirs` options:
 
   ```js
   prettier.format("code", {
     parser: "foo",
-    pluginSearchDirs: ["./dir-with-plugins"],
     plugins: ["./foo-plugin"],
+    pluginSearchDirs: ["./dir-with-plugins"],
```
[the remark this answers is not in the document]
Comment (adamhl8) [reply docs/plugins.md:30]: This is just a change of the ordering of the text to make things a bit more consistent throughout the docs.

File: docs/plugins.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fisker) [reply docs/plugins.md:10]: Do this in a seperate PR.

**Comment by adamhl8** [comment]
Removed changes to the docs as requested and rebased on main.

File: src/common/load-plugins.js [hunk 3]
```diff
@@ -25,11 +26,48 @@ function load(plugins, pluginSearchDirs) {
...
+      ? path.resolve(autoLoadDirectory)
+      : ""; // Make sure autoLoadDir isn't null.
+
+    const isProjectRoot = findProjectRoot(__dirname) === autoLoadDirectory; // Make sure we don't go outside projectRoot.
+
+    // Get the parent directory of autoLoadDir.
+    // It will be the pnpm virtual store if we're not at the project root and if "prettier@" is found.
+    const pnpmVirtualStore =
+      !isProjectRoot &&
+      autoLoadDirectory.includes("prettier@") &&
```
Comment (fisker) [inline src/common/load-plugins.js:58]: `path.dirname(...).startsWith("prettier@")`

File: src/common/load-plugins.js [hunk 4]
```diff
@@ -25,11 +26,48 @@ function load(plugins, pluginSearchDirs) {
...
-    const autoLoadDir = thirdParty.findParentDir(__dirname, "node_modules");
-    if (autoLoadDir) {
-      pluginSearchDirs = [autoLoadDir];
+    // This is likely the project root. In projects using pnpm, this directory is named "prettier@<version>".
+    let autoLoadDirectory = thirdParty.findParentDir(__dirname, "node_modules");
+    autoLoadDirectory = autoLoadDirectory
+      ? path.resolve(autoLoadDirectory)
+      : ""; // Make sure autoLoadDir isn't null.
+
+    const isProjectRoot = findProjectRoot(__dirname) === autoLoadDirectory; // Make sure we don't go outside projectRoot.
```
Comment (fisker) [inline src/common/load-plugins.js:52]: Though it's not recommend, but we support global installation, I think editors use global one when Prettier is not installed locally.

File: src/common/load-plugins.js (hunk 3 above)
Comment (adamhl8) [reply src/common/load-plugins.js:58]: @fisker `pnpmVirtualStore` gets set to the path returned by `path.dirname()` (assuming the first two checks pass). It's not supposed to be `true`. I might be misunderstanding your comment though.

I'll change `autoLoadDirectory.includes("prettier@")` to `autoLoadDirectory.startWith("prettier@")`.

File: src/common/load-plugins.js (hunk 4 above)
Comment (adamhl8) [reply src/common/load-plugins.js:52]: That shouldn't affect global install. `isProjectRoot` is only used to check if we're in the pnpm virtual store (line 57).

File: src/common/load-plugins.js (hunk 3 above)
Comment (fisker) [reply src/common/load-plugins.js]: Isn't it a absolute path? Are we testing the dir name is `pretty@2.5.1`?

Comment (adamhl8) [reply src/common/load-plugins.js]: When using pnpm, the parent directory of `autoLoadDirectory` is the pnpm virtual store. We check to see if `autoLoadDirectory` starts with `prettier@` so we know that pnpm is being used. Then we grab the parent directory with `path.dirname`.

Comment (fisker) [reply src/common/load-plugins.js]: I understand that. `autoLoadDirectory` is a absolute path, `.startsWith()` will never be true, you need get the dirname.

Comment (fisker) [reply src/common/load-plugins.js]: We can know the version of Prettier by reading `package.json` so we can test the dirname is exactly `prettier@<version>`

Comment (adamhl8) [reply src/common/load-plugins.js]: Ah, gotcha, sorry. Pushed a commit that should fix that. Would it not be better/less fragile to just check for `startsWith("prettier@")`. (For example, if pnpm changes how it appends the version number.)

**Comment by adamhl8** [comment]
@fisker Was hoping you could take a look at this again. I fixed the merge conflicts and all tests pass.

**Comment by adamhl8** [comment]
Added proper tests to make codecov happy. Are there any issues that are preventing this from being merged? @fisker

**Comment by fisker** [comment]
Sorry for the delay, I haven't take a deep look. But the plugin related part already changed in `next` branch. Would like to change target base to `next`?

**Comment by adamhl8** [comment]
@fisker I closed this by mistake. Are you able to reopen it and change the base to `next`?

I updated my changes/tests to work with `next`.

**Comment by adamhl8** [comment]
Ah weird. New PR is here: #13583

**Pull request closed without merging**